### PR TITLE
Kubernetes Pod Operator: check pod completion reason if completed during deffered logging interval

### LINKED ISSUE #41867: Kubernetes Pod Operator: race condition when using deferrable and logging_interval
### Apache Airflow Provider(s)

cncf-kubernetes

### Versions of Apache Airflow Providers

apache-airflow-providers-cncf-kubernetes==8.4.1

### Apache Airflow version

2.9.1

### Operating System

Linux

### Deployment

Astronomer

### Deployment details

Originally we noticed this in astronomer in a dag using astronomer-cosmos.  However I have been able to reliably reproduce the issue in airflow running through a python debugger in vscode using just the Kubernetes Pod Operator.

### What happened

When running a DAG we noticed a task that failed according to the logs output from DBT.  However airflow still marked the task as successful and then proceeded with the next task.  

In the airflow logs we see a line like 
```
[2024-08-28, 13:30:10 UTC] {triggerer_job_runner.py:631} INFO - Trigger my_task_group/manual__2024-08-28T11:39:57.188213+00:00/regional_tickets.base_tickets_run/-1/1 (ID 711) fired: TriggerEvent<{'status': 'running', 'last_log_time': DateTime(2024, 8, 28, 13, 29, 1, 820284, tzinfo=Timezone('UTC')), 'namespace': 'sidereal-protostar-2231', 'name': 'my-dag-run-2p3iotfr'}>
```
Then logs from DBT that show task has failed followed by
```
2024-08-28, 13:30:56 UTC] {taskinstance.py:352} INFO - Marking task as SUCCESS. dag_id=my_dag, task_id=regional_tickets.base_tickets_run, run_id=manual__2024-08-28T11:39:57.188213+00:00, execution_date=20240828T113957, start_date=20240828T123001, end_date=20240828T133056
```

### What you think should happen instead

The task should have failed and retried.

### How to reproduce

I have created a simple dag like
```python
from airflow.providers.cncf.kubernetes.operators.pod import KubernetesPodOperator
import airflow
from airflow import DAG
from datetime import timedelta

with DAG(
        'kubernetes_defer_test',
        default_args={
            'start_date': airflow.utils.dates.days_ago(0),
            'retries': 20,
            'retry_delay': timedelta(seconds=5)
        },
        schedule_interval=None,
        catchup=False,
) as dag:

    task1 = KubernetesPodOperator(
        name="hello-world",
        image="python:3.11-slim",
        cmds=['python', '-c'],
        arguments=["import time; time.sleep(5); print('hello'); time.sleep(30); print('world'); import sys; sys.exit(1)"],
        task_id="dummy_run",
        deferrable=True,
        logging_interval=10,
    )
```

And then placing a breakpoint on https://github.com/apache/airflow/blob/e8888fe055ac8c341e2fa6631ff6ac5f089d54c5/airflow/providers/cncf/kubernetes/utils/pod_manager.py#L427 
I leave it wait here until the pod execution fails, and then when it resumes the task will be marked as successful.  

### Anything else

I hope this is a faithful replication of the bug.  I have seen it just by running the dag in an airflow deployment, but it is intermittent until the timing lines up correctly.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by johnhoran**
---
**^ Add meaningful description above**
closes #41867

This is to fix a bug in the Kubernetes pod operator where if the task wakes up from a deferred state to collect logs.  If the task completes during that time then operator won't defer it again, however it also doesn't evaluate the reason for the completion, so if the pod ends in an error, it will still be marked as successful.

The fix here is to always defer when the task resumes while the pod is in a running state, so that the next resume will notice the pod exit and act accordingly, however alternatively we could check the pod exit state and skip the deferment.  I'm happy to amend the PR if this is preferred.

### CODE REVIEW

**Comment by dstandish** [comment]
Yeah I think it's best to avoid the unnecessariy defer / resume cycle.  seems you could refactor the logic under the `if event["status"] in ("error", "failed", "timeout")` block into a method and invoke it in the case you describe. Or something like that.  But, wasteful to go through a another defer-resume cycle.

**Comment by johnhoran** [comment]
@dstandish thanks, I've refactored the code slightly so it will run the success or failure section if the pod completes without deferring again.

**Comment by RNHTTR** [comment]
@johnhoran Can you please fix the failing tests?

**Comment by johnhoran** [comment]
> Can you please fix the failing tests?

@RNHTTR all resolved now, I've also added a test case which would have failed if the failure status wasn't handled.

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 1]
```diff
@@ -391,9 +391,7 @@ def __init__(
         self.skip_on_exit_code = (
             skip_on_exit_code
             if isinstance(skip_on_exit_code, Container)
-            else [skip_on_exit_code]
-            if skip_on_exit_code is not None
-            else []
+            else [skip_on_exit_code] if skip_on_exit_code is not None else []
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/operators/pod.py]: I would like to suggest this instead:
```
    self.skip_on_exit_code = None
    if isinstance(skip_on_exit_code, Container):
        self.skip_on_exit_code = skip_on_exit_code
    elif skip_on_exit_code:
        self.skip_on_exit_code = skip_on_exit_code
```
why:
The multiple  if elses in the comprehension are hard to follow.  And, I think it's nice to use None when there's no value (instead of empty list) and I looked at where this is used and it seems it should work the same.
I think it's nice to use None instead of empty list .

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 2]
```diff
@@ -787,9 +774,27 @@ def trigger_reentry(self, context: Context, event: dict[str, Any]) -> Any:
                     if pod_log_status.running:
                         self.log.info("Container still running; deferring again.")
                         self.invoke_defer_method(pod_log_status.last_log_time)
+                    else:
+                        event = event.copy()
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/operators/pod.py]: don't really like the idea that we're changing the event. the event is what came from the trigger.  there's not really a need to mess with it. yes we have checked if the pod status has changed. but we don't need to use the event to communicate this. the event can be left alone.

Comment (dstandish) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: ok so, here's a suggestion.

what we're dealing with is the situation where the pod happens to complete after the trigger defers and before the task finishes.  

this is a sort of less common scenario.  so, rather than change the PodLoggingStatus object, and modify the event object just read the pod again, and handle accordingly.  

e.g. add read pod here and evaluate success:

if pod is succeeded, do success logic; if pod failed, do failed logic.  i guess you don't even need to pull the success / failure logic out into methods.  you just need to move it down.  you now have two ways to discern success or failure. usually it will come from event; but if event data is stale it will come from the pod read.

Comment (dstandish) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: e.g. something like 
```
succeeded = False
if event == running blah:
    ... other logic
    if pod_log_status.running:
        self.log.info("Container still running; deferring again.")
        self.invoke_defer_method(pod_log_status.last_log_time)
    else:  # now we know pod status is different; let's read the pod again.
        pod = self.read_pod(...)
        succeeded = container_is_succeeded(...)  # <-- bam now you know to do failure logic even though event is success

```

File: airflow/providers/cncf/kubernetes/operators/pod.py (hunk 1 above)
Comment (johnhoran) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: I've amended it.  FYI this was just a change from running black.

File: airflow/providers/cncf/kubernetes/operators/pod.py (hunk 2 above)
Comment (johnhoran) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: Agreed, I've extracted out the pod_status from event so the event doesn't get modifed

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 3]
```diff
@@ -745,6 +746,7 @@ def trigger_reentry(self, context: Context, event: dict[str, Any]) -> Any:
         grab the latest logs and defer back to the trigger again.
         """
         self.pod = None
+        pod_status = event["status"]
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/operators/pod.py]: ok it's subtle but you are still conflating the event "status" with the pod status.   i recommend keeping them separate.  you can either consider each independently (e.g. if pod_status is this, elif event_status is this etc) or you can add a new variable that essentially aggregates the pod status and event status.  you can see why this matters if you look at line 787.   there's no pod status "timeout".  so this could be confusing.

Comment (johnhoran) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: Okay, I've kept them entirely separate.

**Comment by johnhoran** [comment]
Looks like this was resolved by https://github.com/apache/airflow/pull/42815
